Prediction API

cmelone opened the pull request:
This PR implements an endpoint that accepts the following payload via a `GET` request to `/v1/allocation`:

```json
{
    "package": {
        "name": "string",
        "version": "string",
        "variants": "string"
    },
    "compiler": {
        "name": "string",
        "version": "string"
    }
}
```

It will also accept an array of these dictionaries.

The program will then search for 4 ≤ x ≤ 5 samples with the following covariates (in priority order):

1. `("pkg_name", "pkg_version", "compiler_name", "compiler_version")`
2. `("pkg_name", "compiler_name", "compiler_version")`
3. `("pkg_name", "pkg_version", "compiler_name")`
4. `("pkg_name", "compiler_name")`
5. `("pkg_name", "pkg_version")`
6. `("pkg_name",)`

We've chosen this order after running simulations of this prediction algorithm and finding that (1) is the best predictor of resource usage.

After collecting the sample, the program returns a mean of past CPU and memory usage to suggest an appropriate k8s allocation. For the moment, we are only targeting requests.

Additionally, we have implemented a safeguard ensuring that the requests will not be set below what is currently allocated for each package. https://github.com/spack/spack/pull/42351, which was merged a few days ago, increases the allocation for many packages (based on max memory per package, rather than mean as we do here). I am hypothesizing that this will likely render many of the predictions futile, but it's still a good opportunity for a Gantry trial run.

I'm looking for any ways to improve the code or make it more concise!

cmelone committed:
first prediction functionality

cmelone committed:
1

cmelone committed:
add basic routing

cmelone committed:
check whether predictions are lower than current allocation

cmelone committed:
adjust conversions

cmelone committed:
fix some bugs

cmelone committed:
no black current_mapping

cmelone committed:
readability

cmelone committed:
small fixes

alecbcs commented on gantry/views.py:
```diff
@@ -36,3 +37,83 @@ async def collect_job(request: web.Request) -> web.Response:
     return web.Response(status=200)
+
+
+@routes.post("/v1/allocate")
```
```suggestion
@routes.get("/v1/allocation")
```

The endpoint name `allocate` gives me the impression that we're allocating the resources in this request. Not a hard request by any means, but I might suggest changing it instead to `allocation` showing that folks are retrieving an allocation for a package. 

I'd also suggest accepting a `GET` request here instead of `POST`. Typically I've implemented/seen post requests reserved for actively modifying a database or system whereas this endpoint is non-destructive/retrieving data from the system.

cmelone replied:
I was thinking [the same](https://github.com/spack/spack/pull/41622#discussion_r1470591283) 🤪

To make this work, we'd need to instruct the client to pass url-encoded JSON into one of the parameters. 

Here's an example request:

```
{"hash":"feifneifenife","package":{"name":"salmon","version":"1.10.2"},"compiler"{"name":"gcc","version":"7.3.1"}}
```

When encoded, it's around 200 bytes. The current restriction in aiohttp is 8190 bytes. I could increase the limit to something like 100k bytes (or larger, I'm not sure what the upper limit on the amount of specs we'd need to handle here is).

cmelone replied:
@kwryankrattiger do you have an idea of how many specs would get sent to the API at once? I'm assuming it would be the size of the largest stack? 

Trying to think if there is some sort of limit we should have on amount of specs we'll process in a single request.

kwryankrattiger replied:
We can make it one spec per request, that would make GET requests make more sense. GET `v1/dyanmic-mapping?data=<base64encodedjson>`

I would suspect we would be seeing up to a thousand specs per stack for the existing stacks. We are also working on a new internal workflow that prunes before we apply the configs which would reduce that to closer to 10s or 100s on average, but the max would still be high for the rebuild everything case.

cmelone replied:
I'll plan for up to a thousand in a single request, and see how that goes

My instinct is that it makes sense to process as many specs as we can in a single request, as we can eliminate the overhead of thousands of requests at the same time. I don't know if a large request would wipe out these theoretical benefits though

kwryankrattiger replied:
we can play with it, that is why I have the two modes because I wasn't sure if batching or individual requests would be better.

We can start with batch and see how it goes., we will be around a few MB for the larger messages, but if that is too large we can also look at setting a limit on the specs per requests.

alecbcs commented on gantry/views.py:
```diff
@@ -36,3 +37,83 @@ async def collect_job(request: web.Request) -> web.Response:
+    if isinstance(payload, dict):
+        response = await predict_single(request.app["db"], payload)
+    else:
+        response = await predict_bulk(request.app["db"], payload)
```
I like idea of checking the payload type here to initiate a bulk request! Tiny sanity check but is it possible for validate to ever return something that isn't a dict or list? It might be worth explicitly checking if it is a list instead of a dict and then falling back to a dict instead.

cmelone replied:
69c259c1101a38e90c83e67e05d7f2e0c3bb9ade improves the logic of the validate function a bit. It won't return true unless passed a dict or list. I'm going to keep these as individual if statements for now, but will add some tests to check for this.

alecbcs commented on gantry/util/k8s.py:
```diff
@@ -0,0 +1,14 @@
+# these functions convert the predictions to k8s friendly format
+
+
+def convert_cores(cores: float) -> str:
+    """cores to millicores"""
+    return str(int(cores * CORES_TO_MILLICORES)) + "m"
```
Let's talk with @tgamblin, @scottwittenburg, and @jjnesbitt -- It'd be great for understandability of the system if we could get away from using partial CPU values.

jjnesbitt replied:
I don't understand the question, partial CPU values are a core part of the Kuberenetes API.

cmelone replied:
CPU units are now cores rather than millicores and the predictions are rounded up

alecbcs commented on gantry/routes/prediction/prediction.py:
```diff
@@ -0,0 +1,146 @@
+    returns: see predict_single
+    """
+
+    return [await predict_single(db, build) for build in builds]
```
Do you happen to know if this executes in parallel? If not, we could look into asyncio task pools if this gets slow.

cmelone replied:
Great catch! this should work now with [a2aa3db](https://github.com/spack/spack-gantry/pull/5/commits/a2aa3dbfbbf7b7aa5fc0234cf3db0fddcda91916)

alecbcs reviewed (changes requested):
This is looking really good! Just a few initial impressions and a small suggestion.

cmelone committed:
use GET endpoint rather than POST for allocation

cmelone committed:
allocate -> allocation

Co-authored-by: Alec Scott <scott112@llnl.gov>

cmelone committed:
simplify validation logic

cmelone committed:
bulk predictions run in parallel

cmelone committed:
fmt changes

cmelone committed:
cpu units now in cores rather than millicores

cmelone committed:
request limit 8MB

cmelone committed:
add variant filtering

cmelone commented on gantry/routes/prediction/prediction.py:
```diff
@@ -124,6 +127,41 @@ async def get_sample(db: aiosqlite.Connection, build: dict) -> list:
+    # we prefer exact matches but
+    # we can accept the row if all the values of
+    # EXPENSIVE_VARIANTS match
+    for row in sample:
+        row_variants = spec.spec_variants(row[0])
+        if build_variants == row_variants:
+            matched_rows.append(row)
+        elif all(
+            row_variants.get(variant) == build_variants.get(variant)
+            for variant in EXPENSIVE_VARIANTS
+        ):
+            matched_rows.append(row)
```
I'd appreciate some advice on how to incorporate this logic into the sql query. With this setup, we can run into a situation where the db returns the five most recent builds, but none of them match the variants we would like to. For instance, if the 5 most recent builds of Trilinos are all CUDA builds, a request for a ROCm prediction would yield no results, giving the client the default prediction, which is not what we want.

Ideally, we would be able to filter for variants directly within the database, but I'll have to look into the best way to efficiently do this. If anyone has tips or thoughts, that would be great!

cmelone replied:
see https://github.com/spack/spack-gantry/pull/5/commits/c52998ff6f063b5969f306c3ad82ff7251432918

cmelone committed:
updated standard allocations

cmelone committed:
Re-write the variant filtering logic.

Rather than grabbing the 4-5 rows that match based on other params (name, version for both pkg and compiler) and then filtering by variant, these queries are now split up into two. The program will first try to do an exact match on variants. If that doesn't work, it'll try to match based on expensive variants. This increases the amount of queries being done, but is a more robust system and will result in more matches.

cmelone committed:
formatting

cmelone committed:
formatting 😵‍💫

cmelone committed:
logging -> logger usage

cmelone committed:
remove unnecessary comment [ci skip]

alecbcs commented on gantry/routes/prediction/prediction.py:
```diff
@@ -0,0 +1,214 @@
+IDEAL_SAMPLE = 5
+DEFAULT_CPU_REQUEST = 1.0
+DEFAULT_MEM_REQUEST = 2 * 1_000_000_000  # 2GB in bytes
+EXPENSIVE_VARIANTS = {
+    "sycl",
+    "mpi",
+    "rocm",
+    "cuda",
+    "python",
+    "fortran",
+    "openmp",
+    "hdf5",
+}
```
Is there a way we could dynamically build this list by looking at the variant diffs in the db and their corresponding build times/resource utilizations?

cmelone replied:
The way I got these is processing all builds and selecting the variants that have some association with increased memory and cpu usage. We could integrate that logic into here but I think it would be quite expensive, considering that this list is unlikely to change drastically. 

I would prefer either manual updating or with some sort of cron job but happy to consider all options

alecbcs replied:
I think an asynchronous background job is a great idea. I'm okay not requiring that in this PR on that however. Can you create an issue to remind us to write a background job to dynamically build these values before we forget?

cmelone replied:
yup will do!

cmelone committed:
move validate payload for prediction into util for easier testing

cmelone committed:
make a couple changes to prediction to make testing easier

cmelone committed:
remove bulk prediction functionality

we have two options for accepting multiple predictions:

1) instructing the client to send a request for each prediction
2) allowing the client to pass a list of specs to predict for

the second option presents a problem when we get into the weeds
of GET request length limits. since specs can vary widely in
length when running json.dumps() on them, it would be difficult
to either instruct the client to limit their length, or to
somehow validate it.

when we were allowing bulk prediction in one GET request,
it was handled through an asyncio.gather() call. this approach
is approximately 2x faster than individual HTTP requests when
testing 5000 consecutive calls to the API (4s vs 8s). In this case,
8 seconds is not very long so it's not a problem, but we can revisit this
in the future if we run into performance issues.

cmelone committed:
fix: remove unnecessary condition in validate_payload

cmelone committed:
fix: remove custom GET length, 8K is sufficient for one url-encoded json spec

cmelone committed:
fix: remove hash from GET payload

cmelone committed:
fix: update docstring for allocation view

cmelone committed:
allocation API now accepts spec string rather than JSON payload

cmelone committed:
formatting fixes

cmelone committed:
formatting fixes

cmelone committed:
black is not being consistent :(

cmelone committed:
fix: race condition where spec passed by reference was modified by `get_sample`

alecbcs commented on gantry/routes/prediction/prediction.py:
```diff
@@ -0,0 +1,192 @@
+}
+
+
+async def predict_single(db: aiosqlite.Connection, spec: dict) -> dict:
```
```suggestion
async def predict(db: aiosqlite.Connection, spec: dict) -> dict:
```

Since we dropped the multi-spec prediction I'm tempted to say we should rename this to simply be predict.

alecbcs commented on gantry/routes/prediction/prediction.py:
```diff
@@ -0,0 +1,192 @@
+        }
+
+    if os.environ.get("PREDICT_STRATEGY") == "ensure_higher":
+        ensure_higher_pred(predictions, spec["pkg_name"])
```
To make this library less reliant on our specific application use case we should try to instead pass in an optional lower bound as a parameter to predict or use a prediction class to store the information which is instantiated with the strategy/lowerbound instead of making an environment call outside of main.

cmelone replied:
`strategy` is now an optional parameter

cmelone committed:
`predict_single` -> `predict`

Co-authored-by: Alec Scott <scott112@llnl.gov>

cmelone committed:
decouple prediction strategy from env variable

alecbcs reviewed (approved):
Looks good to me. Thanks @cmelone!
